### ISSUE TITLE
Defined refs are unresolved in phpstorm

### ISSUE THREAD

**Issue by LeoDaVinciCodes** [studiometa/js-toolkit#266]
Hi there,

I am testing your toolkit but having a hard time fighting suspicious unresolved ref variables. Maybe you could point me in the right direction?

My component test file looks like this:
```js
import { Base } from '@studiometa/js-toolkit';

export default class Foo extends Base {
  static config = {
    name: 'Foo',
    refs: ['bar']
  };

  mounted() {
    this.$refs.bar;
    console.log('Child mounted');
  }
}
```

My main entry point looks like this:
```js
import { Base, createApp } from '@studiometa/js-toolkit';
import Foo from './Foo.js';

class App extends Base {
  static config = {
    name: 'App',
    components: {
      Foo,
    }
  };
}

export default createApp(App, document.body);
```

The component does indeed mount correctly and the refs are also accessible, but phpstorm keeps complaining about the `bar` ref, which is allegedly undefined.
<img width="1379" alt="CleanShot 2022-08-15 at 04 06 30@2x" src="https://user-images.githubusercontent.com/96996118/184565710-9b2ee497-4f75-41f7-be00-ae99b952b7ff.png">

I also tried to fight the issue by using `jsdoc` statements and included `@studiometa/eslint-config` & `@studiometa/prettier-config` with no success. Do you have an idea what's wrong here on my end? Using @studiometa/js-toolkit@**^2.4.1**

**Comment by titouanmathis** [comment]
Hi @LeoDaVinciCodes, glad to see you trying our toolkit 🙂 

I have not yet found a way to automatically type instance properties based on the given configuration. TypeScript does not play nicely with classes and static properties when using JSDocs. I am still looking for a way to achieve this, either with JSDocs blocks of in TypeScript directly.

The workaround that we use is to redefine an interface in JSDocs comments and use the `@this` block to override the type of `this` in class methods:

```js
import { Base } from '@studiometa/js-toolkit';

/**
 * @typedef {Component & {
 *   $refs: {
 *     btn: HTMLButtonElement,
 *   },
 * }} ComponentInterface
 */

export default class Component extends Base {
  static config = {
    name: 'Component',
    refs: ['btn'],
  };

  /**
   * @this {ComponentInterface}
   */
  mounted() {
    this.$refs.btn; // This should be of type `HTMLButtonElement`
  }
}
```

Hope this helps!

**Comment by LeoDaVinciCodes** [comment]
Hey @titouanmathis, thank you for the fast response. I tried it with the doc statement sadly without success. I am also not using typescript in the example project and are naming my files all with the `.js` extension. When I clone your repo I do not get such conflict warnings. So it seems that it has something to do with my local phpstorm configuraiton, but I am not able to figure out what I need to change. I'd like to use your setup completely and also include your jshint and prettier config, since they make a lot of sense. Do you have any further ideas, where I could have some kind of misconfiguration?

All the best!

**Comment by LeoDaVinciCodes** [comment]
Hey @titouanmathis after updating phpstorm it seems to work with your js doc example as well as with this documentation style for better documenting:

```js
import { Base } from '@studiometa/js-toolkit';

/**
 * @typedef {import('@studiometa/js-toolkit/Base').BaseOptions} BaseOptions
 */

/**
 * @typedef {Object} ComponentRefs
 * @property {HTMLElement} btn the btn to work with an awesome framework
 */

/**
 * @typedef {Object} ComponentPrivateInterface
 * @property {ComponentRefs} $refs
 */

/**
 * @typedef {Component & ComponentPrivateInterface} ComponentInterface
 */

export default class Component extends Base {
  static config = {
    name: 'Component',
    refs: ['btn'],
  };

  /**
   * @this {ComponentInterface}
   */
  mounted() {
    const foo = this.$refs.btn; // This should be of type `HTMLButtonElement` and shows the corresponding definition on hover
  }
}
```

I also cloned your `@studiometa/ui` kit and took a ride and I am also going to use your fantastic `@studiometa/eslint-config` and `@studiometa/prettier-config` since they makes a lot of sense to me. You set up a truly productive and consistent workspace with sensible default configs. Thank you very much for sharing this with the community.

**Comment by titouanmathis** [comment]
Thanks for the feedback @LeoDaVinciCodes, it is well appreciated! 

I am closing this issue, as it seems to work for you now.

**Comment by titouanmathis** [comment]
Hey @LeoDaVinciCodes, we published an update in [v2.4.4](https://github.com/studiometa/js-toolkit/releases/tag/2.4.4) to improve how we can use types with components. We also added a [new page](https://js-toolkit.studiometa.dev/guide/going-further/typing-components.html) to the documentation with some examples both with JSDoc and TypeScript, you should check it out!
